### PR TITLE
Use of official Starknet client

### PR DESCRIPTION
**Pull request by Eagle941**
This PR replaces the hand crafted Starknet RPC client with the official RPC client from crate `starknet_providers`.

Also, the Cairo dependencies are updated to `v2.7.0` and blockifier dependency is updated to use the `VisitedPcs` trait.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -9,8 +9,8 @@ members = [
 [workspace.package]
 # Edition 2021 is chosen to support async.
 edition = "2021"
-# Version chosen only because 1.78.0 is the stable version when writing this Cargo.toml
-# If this version is changed, amend the version set in `build.yml`
+# Version 1.80.1 or higher is not compatible because it enforces check of `cfg` names and values. The dependency `blockifier` currently uses unknown values for `target_pointer_width`.
```
Comment (iamrecursion) [inline Cargo.toml:12]: This has been filed as an issue with blockifier, right?

File: starknet-replay/src/storage/rpc/mod.rs [hunk 2]
```diff
@@ -75,58 +112,30 @@ pub struct RpcStorage {
...
     /// This function queries the number of the most recent Starknet block.
     ///
     /// # Errors
     ///
     /// Returns [`Err`] if the request fails.
-    pub fn starknet_block_number(&self) -> Result<BlockNumber, DatabaseError> {
-        let response = self.send_request("starknet_blockNumber", None)?;
-        let result: u64 = response.result()?;
-        Ok(BlockNumber::new(result))
+    #[allow(clippy::missing_panics_doc)] // To avoid false positives caused by OK() statement.
```
Comment (iamrecursion) [inline starknet-replay/src/storage/rpc/mod.rs:134]: This doesn't make sense to me. How does the `Ok` statement create false positives for panics?

File: starknet-replay/src/storage/rpc/transaction.rs [hunk 3]
```diff
@@ -1,102 +1,299 @@
...
-            "0x1" => Ok(Transaction::Invoke(InvokeTransaction::V1(
-                serde_json::from_value(transaction)?,
-            ))),
-            "0x3" => Ok(Transaction::Invoke(InvokeTransaction::V3(
-                serde_json::from_value(transaction)?,
-            ))),
-            x => Err(DatabaseError::Unknown(format!(
-                "unimplemented invoke version: {x}"
-            ))),
+pub fn convert_transaction(
```
Comment (iamrecursion) [inline starknet-replay/src/storage/rpc/transaction.rs:50]: Okay this method is absolutely mammoth. It needs to be split up, and thankfully the wrapped `tx` in most cases appears to be a different type that could be forwarded to another method.

File: starknet-replay/src/storage/rpc/visited_pcs.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
+use std::collections::hash_map::{Entry, IntoIter, Iter};
+use std::collections::{HashMap, HashSet};
+
+use blockifier::state::state_api::State;
+use blockifier::state::visited_pcs::VisitedPcs;
+use starknet_api::core::ClassHash;
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
```
Comment (iamrecursion) [inline starknet-replay/src/storage/rpc/visited_pcs.rs:8]: This file needs comments.

**Review by iamrecursion — CHANGES_REQUESTED** [review]
Mostly good, but needs some re-engineering in places and some additional documentation.

File: Cargo.toml (hunk 1 above)
Comment (Eagle941) [reply Cargo.toml:12]: They have sorted it out with `sequencer`

**Review by iamrecursion — APPROVED** [review]

**Pull request merged**